<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <link rel="stylesheet" href="style/base.css" type="text/css" />
    <title>Computational Analyses of Light Interacting Genes</title>
    <style>
        .image-container, .video-container {
            display: block;
            margin-left: auto;
            margin-right: auto;
            border-radius: 15px;
            box-shadow: 0 4px 8px rgba(0,0,0,0.2);
            overflow: hidden;
            width: 70%; /* Adjust based on preference */
        }
        
        .video-container {
            background-color: #f0f0f0; /* Light grey, adjust color as needed */
            padding: 15px;
        }
        
        img {
            width: 100%; /* Makes the image responsive to the container size */
            height: auto;
        }
        
        iframe {
            display: block;
            width: 100%;
            height: 315px; /* You can adjust this */
        }
    </style>
</head>
<body>
    <div class="document">
        <h1 style="text-align:center;">Computational Analyses of Light Interacting Genes</h1>
        <hr>


<div class="tool-description">
    <strong>1. Phylogenetically Informed Annotation (PIA) of the Light Interaction Tool Kit (LIT)</strong>
    <p>PIA (Phylogenetically-Informed Annotation) finds Light Interacting Genes in transcriptomes or new genomes by efficiently integrating uncharacterized genes into pre-calculated phylogenies of gene families. For more details, refer to the <a href="https://link.springer.com/article/10.1186/s12859-014-0350-x" target="_blank">published article</a>. TOOL CURRENTLY UNDER RE-CONSTRUCTION</p>
</div>
        <hr>
<div class="tool-description">
    <strong>2. Opsin Phenotyping Tool for Inferring Color Sensitivity (OPTICS)</strong>
    <p>OPTICS employs machine learning models to accurately predict the lambda-max phenotype of opsins from gene sequences alone. For more details on the database and model training procedures, see the <a href="https://www.biorxiv.org/content/10.1101/2024.02.12.579993" target="_blank">related preprint</a>.</p>
    <p><a href="https://tinyurl.com/nm7tt3k8" target="_blank">Here</a> is a link to a bibliography of the publications used to build VPOD_1.0.</p>
    <p>If you know of publications for training opsin ML models not included in the VPOD_1.0 database, please send them to us through <a href="https://tinyurl.com/29afaxyr" target="_blank">this form</a>.</p>
</div>
        <hr>
        <p>Provided by Oakley Lab UCSB. Funded by NSF and UCSB's EEMB Department. Additional support and hardware provided by <a href="http://csc.cnsi.ucsb.edu">Center for Scientific Computing</a> at UCSB.</p>
        <hr>
        <div class="image-container">
            <img src="images/ring.jpg" alt="Ring Nebula" />
        </div>
        <div class="video-container">
            <iframe src="//www.youtube.com/embed/m23KxhxhTAE?rel=0" frameborder="0" allowfullscreen></iframe>
        </div>
        <a target="_blank" class="reference" href="https://bytebucket.org/osiris_phylogenetics/pia/raw/05d213dcb31e4cda9ccc68201ee7f8854161b2c9/docs/PIAwebManual_090914.pdf">Written PIA manual linked here</a>
        <hr>
        <p>Galaxy is an open, web-based platform for data intensive biomedical research. The Galaxy team is a part of BX at Penn State, and the Biology and Mathematics and Computer Science departments at Emory University. The Galaxy Project is supported in part by NHGRI, NSF, The Huck Institutes of the Life Sciences, The Institute for CyberScience at Penn State, and Emory University.</p>
    </div>
</body>
</html>
